keps/sig-node/1948-deallocate-dp-api/README.md

```diff
@@ -0,0 +1,854 @@
+cleanup, driver unloading, and any other cleanup actions that may be needed.
+
+Since both additions are optional, existing device plugins should continue
+functioning properly with no needed modifications. Only device plugins that wish
+to utilize the new API calls will need to be modified.
+
+### Risks and Mitigations
+
+Only risk is breaking existing device plugins by introducing non optional
+changes. Can be mitigated by enough test coverage.
+
+## Design Details
+
+- Move `PreStartContainer` in `DeviceManager` to be used as a container
+lifecycle hook. (This change isn't truly required, but useful for compatibility
+and organization with the next steps).
+- Add `PostStopContainer` and `Deallocate` calls in the DevicePlugin API.
+- Add `PostStopContainer` as a container lifecycle hook.
+- Add `Deallocate` calls in container manager, taking care to only do so for
+devices that are no longer in the reuse list.
+- Add and modify test cases for both calls.
+- Test with existing device plugins to ensure the changes are non-breaking.
+- Test with new device plugins utilizing such changes to ensure the changes are
+working.
+
+### Test Plan
+
+- Unit tests will be updated to include the new API calls
+- E2E tests should be added with a sample device plugin for verification.
+
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+
+- Gather feedback from developers and surveys, specially about the device reuse
+and possible alternatives.
+- Complete implementation for the new API calls.
+- Tests (unit, e2e) are in Testgrid and linked in KEP.
+
+#### Beta -> GA Graduation
+
+- More rigorous testing as needed/discussed by developers.
+- Larger scale use/testing by interested users with no reported major bugs.
+
+### Upgrade / Downgrade Strategy
+
+As part of the device plugin API, this will follow the same API versioning
+system. This means that it is up to an application (a device plugin) to choose
+the required API version it wants. As long as the cluster has a recent enough
+(to include the required API version) kubernetes, upgrade or downgrades require
+no cluster modifications at all, and are decided on the application level.
+
+### Version Skew Strategy
+
+As part of the device plugin API, this will follow the same API versioning
+system. This means that it is up to an application (a device plugin) to choose
+the required API version it wants. No version skew issues will arise.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+#### How can this feature be enabled / disabled in a live cluster?**
+
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: `DeviceManagerDeallocate`
+  - Components depending on the feature gate: `kubelet`
+
+#### Does enabling the feature change any default behavior?**
+
+No. The feature is optional, no default behavior will change.
+Feature Gate must be enabled for this new functionality.
```

Comment:
The intent of this question is whether there is a change in the default behavior *when the feature is enabled*. That is, does the simple act of enabling the feature change the behavior of some cluster components?